### PR TITLE
fix pulsar test instability

### LINKED ISSUE #8537: Pulsar test instability
Error is occurring to Pulsar test case randomly on multiple of my PRs
```
PulsarConsumerTest.testPartitionLevelConsumerBatchMessages:246 expected [500] but found [491]
```

<details>
  <summary>CI Logs. Click to expand</summary>

  ```
  Error:  Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 15.324 s <<< FAILURE! - in org.apache.pinot.plugin.stream.pulsar.PulsarConsumerTest
  Error:  org.apache.pinot.plugin.stream.pulsar.PulsarConsumerTest.testPartitionLevelConsumer  Time elapsed: 0.845 s
  Error:  org.apache.pinot.plugin.stream.pulsar.PulsarConsumerTest.testPartitionLevelConsumerBatchMessages  Time elapsed: 0.48 s  <<< FAILURE!
  java.lang.AssertionError: expected [500] but found [491]
	  at org.testng.Assert.fail(Assert.java:93)
	  at org.testng.Assert.failNotEquals(Assert.java:512)
	  at org.testng.Assert.assertEqualsImpl(Assert.java:134)
	  at org.testng.Assert.assertEquals(Assert.java:115)
	  at org.testng.Assert.assertEquals(Assert.java:388)
	  at org.testng.Assert.assertEquals(Assert.java:398)
	  at org.apache.pinot.plugin.stream.pulsar.PulsarConsumerTest.testPartitionLevelConsumerBatchMessages(PulsarConsumerTest.java:246)
	  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	  at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	  at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	  at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:108)
	  at org.testng.internal.Invoker.invokeMethod(Invoker.java:661)
	  at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:869)
	  at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1193)
	  at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:126)
	  at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	  at org.testng.TestRunner.privateRun(TestRunner.java:744)
	  at org.testng.TestRunner.run(TestRunner.java:602)
	  at org.testng.SuiteRunner.runTest(SuiteRunner.java:380)
	  at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:375)
	  at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	  at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	  at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	  at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	  at org.testng.TestNG.runSuitesSequentially(TestNG.java:1301)
	  at org.testng.TestNG.runSuitesLocally(TestNG.java:1226)
	  at org.testng.TestNG.runSuites(TestNG.java:1144)
	  at org.testng.TestNG.run(TestNG.java:1115)
	  at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:136)
	  at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeSingleClass(TestNGDirectoryTestSuite.java:112)
	  at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:99)
	  at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:145)
	  at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
	  at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	  at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
	  at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
  
  [INFO] 
  [INFO] Results:
  [INFO] 
  Error:  Failures: 
  Error:    PulsarConsumerTest.testPartitionLevelConsumerBatchMessages:246 expected [500] but found [491]
  ```

</details>

### PR DESCRIPTION
**Pull request by walterddr**
This fixes #8537

### CODE REVIEW

File: pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/test/java/org/apache/pinot/plugin/stream/pulsar/PulsarConsumerTest.java [hunk 1]
```diff
@@ -273,13 +269,31 @@ public void testPartitionLevelConsumerBatchMessages()
...
 
-  public MessageId getMessageIdForPartitionAndIndex(int partitionNum, int index) {
+  private static MessageBatch consumeMessageBatch(PartitionGroupConsumer consumer, MessageId startMsgId,
+      MessageId endMsgId, int expectedMsgCount)
+      throws TimeoutException {
+    int retryCount = 0;
+    MessageBatch messageBatch = null;
+    while (retryCount < DEFAULT_RETRY_COUNT) {
+      retryCount++;
+      messageBatch = consumer.fetchMessages(new MessageIdStreamOffset(startMsgId),
```
Comment (xiangfu0) [inline pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/test/java/org/apache/pinot/plugin/stream/pulsar/PulsarConsumerTest.java:279]: do we know why this call may give fewer records? Is it a bug or something like messages not flushed entirely and available to consume or some messages are actually missing

Comment (walterddr) [reply pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/test/java/org/apache/pinot/plugin/stream/pulsar/PulsarConsumerTest.java:279]: I dunno why this fixes the bug. My guess is that pulsar flush is async but I am not sure. 

GitHub Actions sometimes catch interesting bug b/c the machines are much less powerful

Comment (walterddr) [reply pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/test/java/org/apache/pinot/plugin/stream/pulsar/PulsarConsumerTest.java:279]: I saw the publishRecordBatch was changed in #8017 - any chance @KKcorps can give some insights?

Comment (KKcorps) [reply pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/test/java/org/apache/pinot/plugin/stream/pulsar/PulsarConsumerTest.java:279]: This method was added to handle the scenarios where messages are published in batches which actually changes the format of MessageIds in Pulsar.

The only way to publish in batch is to enable async producer. In sync mode, all messages are flushed as soon as they are received.

**Review by KKcorps — APPROVED** [review]

File: pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/test/java/org/apache/pinot/plugin/stream/pulsar/PulsarConsumerTest.java (hunk 1 above)
Comment (KKcorps) [reply pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/test/java/org/apache/pinot/plugin/stream/pulsar/PulsarConsumerTest.java:279]: @xiangfu0 your suggestion also works. I can actually add a method like `waitForAllRecordsToFlush` and only then exit the setup.

Comment (walterddr) [reply pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/test/java/org/apache/pinot/plugin/stream/pulsar/PulsarConsumerTest.java:279]: according to the documentation https://pulsar.apache.org/api/client/org/apache/pulsar/client/api/Producer.html 
- producer.flush() should wait until all the message finish publishing before returning (or throw exception); thus we can't check on the producer side
- only way to check and wait for all records reaches pulsar for consumption is to actually consume -- which is exactly what this test is doing. 
 
so I guess we should do is wrap the test with 
```
TestUtils.waitForCondition(aVoid -> { 
  // consume and assert 
}, timeoutMs, "")
```

sounds good?

Comment (KKcorps) [reply pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/test/java/org/apache/pinot/plugin/stream/pulsar/PulsarConsumerTest.java]: There is another approach - https://github.com/apache/pinot/pull/8554

**Comment by walterddr** [comment]
it still fails. I wonder if theres' anything wrong with the consumer itself that it didn't pull enough data out. 

alternatively I can still make the wait for condition checker in the test function too so that we can at least get better against the flakiness

**Comment by walterddr** [comment]
This doesn't work. It seems like not a flakiness due to delay async produce.

**Pull request closed without merging**